romancal/flux/flux_step.py

```diff
@@ -0,0 +1,114 @@
+        provided instead of an ASDF filename. Multiple files can be processed via
+        either an association file or wrapped by a
+        `~romancal.datamodels.container.ModelContainer`.
+
+    Returns
+    -------
+    output_models : `roman_datamodels.datamodels.DataModel`, or `~romancal.datamodels.container.ModelContainer`
+        The models with flux applied.
+
+
+    Notes
+    -----
+    Currently, the correction is done in-place; the inputs are directly modified if in-memory DataModels are input.
+    """  # noqa: E501
+
+    spec = """
+    """  # noqa: E501
+
+    reference_file_types = []
+
+    def process(self, input):
+        if isinstance(input, datamodels.DataModel):
+            input_models = ModelContainer([input])
+            single_model = True
+        elif isinstance(input, str):
+            # either a single asdf filename or an association filename
+            try:
+                # association filename
+                input_models = ModelContainer(input)
+                single_model = False
+            except Exception:
+                # single ASDF filename
+                input_models = ModelContainer([input])
+                single_model = True
+        elif isinstance(input, ModelContainer):
+            input_models = input
+            single_model = False
+        else:
+            raise TypeError(
+                "Input must be an ASN filename, a ModelContainer, "
+                "a single ASDF filename, or a single Roman DataModel."
+            )
+
+        for model in input_models:
+            apply_flux_correction(model)
+
+        if single_model:
+            return input_models[0]
+        return input_models
+
+
+def apply_flux_correction(model):
+    """Apply the flux correction
+
+    The input model is expected to be Roman ImageModel-like.
+    The photometry information is taken from `meta.photometry`.
+
+    The model is converted in-place.
+
+    Parameters
+    ----------
+    model : `ImageModel`
+        The model to apply the flux correction to. The model is modified in-place.
+
+    Notes
+    -----
+    The modifications to the model can result in validation issues due to change of units.
+    """
+    # Define the various arrays to be converted.
+    DATA = ("data", "err")
+    VARIANCES = ("var_rnoise", "var_poisson", "var_flat")
+
+    # Check for units. Must be election/second. Otherwise, it is unknown how to
+    # convert.
+    if model.data.unit != u.electron / u.s:
+        log.debug(
+            "Input data is not in units of e/s. Flux correction will not be done."
+        )
+        log.debug("Input data units are %s", model.data.unit)
+        return
+
+    # Apply the correction
+    # Assignments into the model are done through `_instance` to avoid
+    # validation errors on the units.
+    log.debug("Flux correction being applied")
+    c_mj = model.meta.photometry.conversion_megajanskys
+    for data in DATA:
+        model[data] = model[data] * c_mj
```

Comment:
And, sorry, I'm confused all the sudden.  It looks to me like c_mj has units of MJy/sr, and model[data] should have units of something like e/s, so are the units of the new model_data e/s * MJy/sr?  Where are the old units being stripped?

Reply:
The L3 images need to be in MJy / sr.

conversion_megajansky "should" be in (MJy / sr) / (e / s), so that the (e / s) cancels out in that product.  But it looks like we've chosen to define it as the number of MJy/sr producing 1 count per second, eliding the denominator.  I think the easiest thing would be to define c_mj = conversion_megajansky / (1.0 * e / s).  That will need to change to c_mj = conversion_megajansky / (1.0 * DN / s) when the L2 unit changes go in, but that doesn't sound bad.

We will also need to work with RTB to make sure that the photometry tables are consistent with the unit change.  In principle I think we need new files with all of the zero points modified by a factor of the average gain, though in practice I don't think we're presently looking very hard at the overall zeropoints and so if we didn't do this we wouldn't notice.  @tddesjardins